File: Data Structures & Algorithms/jump-game-ii/submission-2.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        dp = [float("inf")]*len(nums)
        dp[-1] = 0

        for i in range(len(nums)-2, -1, -1):
            dp[i] = 1 + min(dp[min(len(nums)-1, i+n)] for n in range(nums[i]+1))

        return dp[0]

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res=nums[0];
        int curmin=1;
        int curmax=1;
        for(int n:nums){
            int tmp = curmin * n;
            curmin = min(min(curmax*n, tmp), n);
            curmax = max(max(curmax*n, tmp), n);
            res = max(res, curmax);
        }
        return res;
    }
};
